src/V3ThreadPool.cpp

```diff
@@ -14,139 +14,71 @@
-    }
-    while (!m_workers.empty()) {
-        m_workers.front().join();
-        m_workers.pop_front();
-    }
-    if (n > 1) {
-        V3LockGuard lock{m_mutex};
-        // Start new threads
-        m_shutdown = false;
-        for (unsigned int i = 1; i < n; ++i) {
-            m_workers.emplace_back(&V3ThreadPool::startWorker, this, i);
-        }
     }
+    m_cv.notify_all();
+    wait();
 }
 
-void V3ThreadPool::suspendMultithreading() VL_MT_SAFE VL_EXCLUDES(m_mutex)
-    VL_EXCLUDES(m_stoppedJobsMutex) {
-    V3LockGuard stoppedJobsLock{m_stoppedJobsMutex};
-    if (!m_workers.empty()) stopOtherThreads();
-
-    if (!m_mutex.try_lock()) {
-        v3fatal("Tried to suspend thread pool when other thread uses it.");
+void V3ThreadPool::enqueue(std::function<void()>&& f) {
+    if (m_workers.empty()) {
+        f();
+    } else {
+        {
+            const V3LockGuard lock{m_mutex};
+            m_queue.push(std::move(f));
+        }
+        m_pendingJobs.fetch_add(1, std::memory_order_release);
+        m_cv.notify_one();
     }
-    V3LockGuard lock{m_mutex, std::adopt_lock_t{}};
-
-    UASSERT(m_queue.empty(), "Thread pool has pending jobs");
-    UASSERT(m_jobsInProgress == 0, "Thread pool has jobs in progress");
-    m_exclusiveAccess = true;
-    m_multithreadingSuspended = true;
 }
 
-void V3ThreadPool::resumeMultithreading() VL_MT_SAFE VL_EXCLUDES(m_mutex)
-    VL_EXCLUDES(m_stoppedJobsMutex) {
-    if (!m_mutex.try_lock()) v3fatal("Tried to resume thread pool when other thread uses it.");
-    {
-        V3LockGuard lock{m_mutex, std::adopt_lock_t{}};
-        UASSERT(m_multithreadingSuspended, "Multithreading is not suspended");
-        m_multithreadingSuspended = false;
-        m_exclusiveAccess = false;
+void V3ThreadPool::wait() {
+    while (m_pendingJobs.load(std::memory_order_acquire) > 0 && !m_shutdown) {
+        std::this_thread::yield();
     }
-    if (!m_workers.empty()) {
-        V3LockGuard stoppedJobsLock{m_stoppedJobsMutex};
-        resumeOtherThreads();
+    if (m_shutdown) {
+        for (auto& worker : m_workers) { worker.join(); }
     }
 }
 
-void V3ThreadPool::startWorker(V3ThreadPool* selfThreadp, int id) VL_MT_SAFE {
-    selfThreadp->workerJobLoop(id);
-}
+void V3ThreadPool::startWorker(V3ThreadPool* selfThreadp) { selfThreadp->workerJobLoop(); }
 
-void V3ThreadPool::workerJobLoop(int id) VL_MT_SAFE {
+void V3ThreadPool::workerJobLoop() {
     while (true) {
-        // Wait for a notification
-        waitIfStopRequested();
-        VAnyPackagedTask job;
+        std::function<void()> job;
         {
-            V3LockGuard lock(m_mutex);
-            m_cv.wait(m_mutex, [&]() VL_REQUIRES(m_mutex) {
-                return !m_queue.empty() || m_shutdown || m_stopRequested;
-            });
+            const V3LockGuard lock{m_mutex};
+            m_cv.wait(m_mutex,
+                      [&]() VL_REQUIRES(m_mutex) { return !m_queue.empty() || m_shutdown; });
```

Comment:
This might be ok, but please double check works right regardless of other threads entering this code between the lock and wait, or within the wait.

Reply:
To rule out the situation when a thread would miss a notification we take locks around every conditional_variable related code. This ensures that threads would not miss it. The condvar condition is also checked under a lock (validated by thread-safety).

I think the problematical situation may occur when a thread waits on a lock when the notification is issued. However,  the `wait` function first checks the condition before waiting for the notification, thus no problem with missed notification here.

Reply:
Ok, please add some comments there or somewhere else appropriate about this, thanks.